src/lib.rs

```diff
@@ -662,57 +647,87 @@ where
         })
     }
 
+    // Main grammar.
     let mut grammar = grammar! {
-        // Lexical grammar.
-        Whitespace = {
-            {{
-                " " | "\t" | "\n" | "\r" |
-                { "//" {{ {!"\n"} .. }*} "\n" } |
-                { "/*" {{ {!"*/"} .. }*} "*/" }
-            }*}
-            {!" "} {!"\t"} {!"\n"} {!"\r"} {!"//"} {!"/*"}
-        };
-        Shebang = { "#!" {{ {!"\n"} .. }*} "\n" };
-
-        IdentStart = {'a'..='z'} | {'A'..='Z'} | "_";
-        IdentCont = IdentStart | {'0'..='9'};
-        NotIdent = { {!'a'..='z'} {!'A'..='Z'} {!"_"} {!'0'..='9'} };
-        Ident = { IdentStart {IdentCont*} NotIdent };
-
-        StrLit = { "\"" {{ { {!"\\"} {!"\""} .. } | { "\\" Escape } }*} "\"" };
-        CharLit = { "'" { { {!"\\"} {!"'"} .. } | { "\\" Escape } } "'" };
-        Escape = "t" | "n" | "r" | "\\" | "'" | "\"";
+        Grammar = { FileStart {rules:{RuleDef*}} FileEnd };
+        RuleDef = { {name:Ident} "=" {rule:Or} ";" };
+        Or = {{"|"?} {rules:{Concat+ % "|"}}};
+        Concat = {rules:{Rule+}};
+        Rule = { {{ {field:Ident} ":" }?} {rule:Primary} {{modifier:Modifier}?} };
+        Primary =
+            {Eat:Pattern} |
+            {Call:Ident} |
+            {Group:{ "{" {{or:Or}?} "}" }};
+        Modifier =
+            {Opt:"?"} |
+            {Repeat:{ {repeat:Repeat} {{ {kind:SepKind} {sep:Primary} }?} }};
+        Repeat =
+            {Many:"*"} |
+            {More:"+"};
+        SepKind =
+            {Simple:"%"} |
+            // HACK(eddyb) should be "%%", but `proc_macro` doesn't preserve
```

Comment:
I'd appreciate if this comment can be clarified not to blame `proc_macro` as I showed over in https://github.com/rust-lang/gll/issues/110#issuecomment-492285319. The issue is steps in-between, including (but potentially not limited to) `<TokenStream as Display>::fmt` and `macro_rules!`.